src/Database/LSMTree/Internal/Merge.hs

```diff
@@ -0,0 +1,354 @@
+        FullEntry (combine mrg (Mupdate v1) (toFullEntry e2))
+      _ ->
+        -- otherwise, just take the first one (OverflowEntry can't be Mupdate!)
+        e1
+
+-- | This function should be called when discarding a 'Merge' before it
+-- was done (i.e. returned 'MergeComplete'). This removes the incomplete files
+-- created for the new run so far and avoids leaking file handles.
+--
+-- Once it has been called, do not use the 'Merge' any more!
+close ::
+     HasFS IO h
+  -> Merge (FS.Handle h)
+  -> IO ()
+close fs Merge {..} = do
+    MRun.close fs mergeWriter
+    closeReaders fs mergeReaders
+
+{-------------------------------------------------------------------------------
+  Read
+-------------------------------------------------------------------------------}
+
+-- | Abstraction for the collection of 'Reader', yielding elements in order.
+-- More precisely, that mean first ordered by their key, then by the input
+-- run they came from. This is important for resolving multiple entries with the
+-- same key into one.
+--
+-- Construct with 'newReaders', then keep calling 'getFromReaders'.
+-- If aborting early, remember to call 'closeReaders'!
+data Readers fhandle = Readers {
+      readersHeap     :: !(Heap.MutableHeap RealWorld (ReadCtx fhandle))
+      -- | Since there is always one reader outside of the heap, we need to
+      -- store it separately.
+    , readersLastRead :: !(IORef (Reader fhandle, ReaderNumber))
+    }
+
+newtype ReaderNumber = ReaderNumber Int
+  deriving (Eq, Ord)
+
+-- | Each heap element needs some more context than just the reader.
+-- E.g. the 'Eq' instance we need to be able to access the first key to be read
+-- in a pure way.
+--
+-- TODO(optimisation): We allocate this record for each k/op. This might be
+-- avoidable, see ideas below.
+data ReadCtx fhandle = ReadCtx {
+      -- We could avoid this using a more specialised mutable heap with separate
+      -- arrays for keys and values (or even each of their components).
+      -- Using an 'STRef' could avoid reallocating the record for every entry,
+      -- but that might not be straightforward to integrate with the heap.
+      readCtxHeadKey   :: !SerialisedKey
+    , readCtxHeadEntry :: !(PrefixEntry fhandle)
+      -- We could get rid of this by making 'LoserTree' stable! (for which there
+      -- is a prototype already)>
+    , readCtxNumber    :: !ReaderNumber
+    , readCtxReader    :: !(Reader fhandle)
+    }
+
+instance Eq (ReadCtx fhandle) where
+  (==) = (==) `on` (\r -> (readCtxHeadKey r, readCtxNumber r))
+
+-- | Makes sure we resolve entries in the right order.
+instance Ord (ReadCtx fhandle) where
+  compare = compare `on` (\r -> (readCtxHeadKey r, readCtxNumber r))
+
+newReaders ::
+     HasFS IO h
+  -> HasBufFS IO h
+  -> [Run (FS.Handle h)]
+  -> IO (Maybe (Readers (FS.Handle h), SerialisedKey, PrefixEntry (FS.Handle h)))
+newReaders fs bfs runs = do
+    readers <- zipWithM (fromRun . ReaderNumber) [1..] runs
```

Comment:
Yes, I should leave a comment.

Reply:
I was thinking about this ordering, and it might be something we can impose later via the "unique run naming scheme" that we're going to need to get non-conflicting names for the run files. One probable option is that we'll have a session-wide counter, which gets incremented each time a new run is created, and we use the counter value for the run names. This counter could be used for the run's `RunFsPaths`, and then we can use that number to determine the age (and therefore the required ordering in the list above) of the input runs

Reply:
Yes, if we decide to have such an invariant on the run number in `RunFsPaths`, that can replace the `ReaderNumber`. The LSM structure will know the run order independently of that, but we could for example assert here that the run numbers match the order of the list.
I don't want to assume such an invariant now, but I'll also note the idea in the comment.